src/CoreWCF.Primitives/src/CoreWCF/Security/SecuritySessionServerSettings.cs

```diff
@@ -0,0 +1,2526 @@
+            }
+
+            internal void TraceSessionClosedFailure(Exception e)
+            {
+                //  SecurityTraceRecordHelper.TraceSessionCloseSendFailure(this.currentSessionToken, GetLocalUri(), e);
+            }
+
+            // SendCloseResponse closes the message and underlying context if the operation completes successfully
+            protected void SendCloseResponse(RequestContext requestContext, Message closeResponse, CancellationToken token)
+            {
+                try
+                {
+                    using (closeResponse)
+                    {
+                        SendMessage(requestContext, closeResponse, token);
+                    }
+                    
+                    TraceSessionClosedResponseSuccess();
+                }
+                catch (CommunicationException e)
+                {
+                    TraceSessionClosedResponseFailure(e);
+                }
+                catch (TimeoutException e)
+                {
+                    TraceSessionClosedResponseFailure(e);
+                }
+            }
+
+            internal Message CreateCloseMessage(CancellationToken token)
+            {
+                RequestSecurityToken rst = new RequestSecurityToken(this.Settings.SecurityStandardsManager);
+                rst.RequestType = this.Settings.SecurityStandardsManager.TrustDriver.RequestTypeClose;
+                rst.CloseTarget = this.Settings.IssuedSecurityTokenParameters.CreateKeyIdentifierClause(this.currentSessionToken, SecurityTokenReferenceStyle.External);
+                rst.MakeReadOnly();
+                Message closeMessage = Message.CreateMessage(this.messageVersion, ActionHeader.Create(this.Settings.SecurityStandardsManager.SecureConversationDriver.CloseAction, this.messageVersion.Addressing), rst);
+                RequestReplyCorrelator.PrepareRequest(closeMessage);
+                if (this.LocalAddress != null)
+                {
+                    closeMessage.Headers.ReplyTo = this.LocalAddress;
+                }
+                else
+                {
+                    if (closeMessage.Version.Addressing == AddressingVersion.WSAddressing10)
+                    {
+                        closeMessage.Headers.ReplyTo = null;
+                    }
+                    else if (closeMessage.Version.Addressing == AddressingVersion.WSAddressingAugust2004)
+                    {
+                        closeMessage.Headers.ReplyTo = EndpointAddress.AnonymousAddress;
+                    }
+                    else
+                    {
+                        throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
+                            new ProtocolException(SR.Format(SR.AddressingVersionNotSupported, closeMessage.Version.Addressing)));
+                    }
+                }
+                this.securityProtocol.SecureOutgoingMessageAsync(closeMessage, token);
```

Comment:
You're calling an async method here. I had a look at SecureOutgoingMessageAsync and it looks like it shouldn't be an async method from my brief check. Can you verify, and either make sure all callers of SecureOutgoingMessageAsync are async and await-ing the the method, or change the method to not be async.

Reply:
made it sync...